Using Github Actions to automatically deploy documentation

louisdewar opened the pull request:
Hi, I thought I'd have a crack at #444 which is to create a Github actions workflow to deploy the documentation on every push to main.

When I was testing I added my development branch `docs-deployment` to the list of branches the the workflow was watching and you can see the automatic deployment here: [https://louisdewar.github.io/ruma/](https://louisdewar.github.io/ruma/) (this branch is now removed from the workflow, only `main` is listed currently).

Right now the flow is push to `main` then this builds the docs and then deploys to a branch called `docs`. In the settings for this repo you can go to `options` -> `GitHub Pages` and then specify the `Source` as the `docs` branch and then specify the root for the folder.
It can take a bit of time but the docs should be visible at https://ruma.github.io/ruma/ once that's complete. If you instead want the URL to be `docs.ruma.io` then that's possible too. I believe it will also be visible at https://ruma.io/ruma because of the way Github pages works but I'm not fully sure.

In https://github.com/matrix-org/matrix-rust-sdk/blob/master/.github/workflows/docs.yml the docs were built for every push to main and for pull requests but only deployed for pushes to main. Is that the behaviour you want to replicate? Currently it does nothing on pull requests but I assume it would be useful for the CI to detect if there are any errors in building docs.

louisdewar committed:
Initial test

louisdewar committed:
Add override to make nightly the default

louisdewar committed:
Remove reference to development branch in workflow

zecakeh commented:
You probably want to exclude `xtask`, because it's an internal package.

I was also wondering if some features need to be activated for the docs to be more complete: `compat`, `unstable-pre-spec` and `unstable-synapse-quirks`.

zecakeh commented on .github/workflows/docs.yaml:
```diff
@@ -0,0 +1,29 @@
+      with:
+        command: doc
+        args: --no-deps --workspace
+    - name: Deploy to gh-pages branch
```
```suggestion
    - name: Deploy to docs branch
```

louisdewar committed:
Fix typo in branch name

louisdewar committed:
Exclude xtask and add features

louisdewar committed:
Removed development branch from workflow

louisdewar commented:
Hi @zecakeh, thanks for the review. I've made the changes you requested:
- Removed `xtask`
- Build docs for features: [`compat`, `unstable-pre-spec`, `unstable-synapse-quirks`] - as an example https://louisdewar.github.io/ruma/ruma_client_api/r0/keys/index.html now shows `upload_signatures` and `upload_signing_keys` which are both behind the `unstable-pre-spec` feature flag and weren't shown before.
- Fixed typo in branch name

I was trying to look into how docs.rs manages to get a nice UI showing which items are dependent on features but I don't think it's built into rustdoc unfortunately.

jplatte commented on .github/workflows/docs.yaml:
```diff
@@ -0,0 +1,29 @@
+    runs-on: ubuntu-latest
+    steps:
+    - name: Checkout repo
+      uses: actions/checkout@master
```
```suggestion
      uses: actions/checkout@v2
```

louisdewar replied:
Done

zecakeh commented:
You're talking about the feature flags list ? Yeah I believe it's only on docs.rs.

But there's a way to show that a feature needs to be enabled for a function or struct. It's used on matrix-rust-sdk: https://github.com/matrix-org/matrix-rust-sdk/blob/ff683602f250e90e83acbbd580252fa8ce6c541a/matrix_sdk/src/client.rs#L898

Renders like this: https://matrix-org.github.io/matrix-rust-sdk/matrix_sdk/struct.Client.html#method.login_with_sso

It's too bad there doesn't seem to be a way to automate this... It might be too much work for this PR.

zecakeh commented:
I found this crate that automates this (you use one attribute instead of two): https://crates.io/crates/doc-cfg

It doesn't seem very popular though.

jplatte commented:
We already use `#[doc(cfg)]` in a few places. A separate PR adding it to unstable modules would be welcome.

jplatte commented on .github/workflows/docs.yaml:
```diff
@@ -0,0 +1,29 @@
+       RUSTDOCFLAGS: "--enable-index-page -Zunstable-options"
+      with:
+        command: doc
+        args: --no-deps --workspace --exclude xtask --features compat,unstable-pre-spec,unstable-synapse-quirks
```
One more cargo option we should use: [`-Zrustdoc-map`](https://doc.rust-lang.org/nightly/cargo/reference/unstable.html#rustdoc-map). And `.cargo/config` has to be updated accordingly, but that's super easy.

jplatte replied:
And for the crate features, I think this is the list we should use: `api,events,signatures,appservice-api,client-api,federation-api,identity-service-api,push-gateway-api,either,rand,markdown,compat,unstable-pre-spec`. Most of these are for the `ruma` crate; `--all-features` would be preferable but can't be used right now due to the TLS crate features of `ruma-client` being mutually exclusive.

zecakeh replied:
We could define a `docs` feature that enables the necessary features, like in matrix-rust-sdk. It's easier to use locally in the command line and can be reused to launch all tests.

jplatte replied:
I don't think that's necessary. I'm planning to have `--all-features` work for ruma-client at some point.

louisdewar replied:
The rustdoc-map flag and new features are all added now

louisdewar committed:
Added new features and rustdoc-map

louisdewar committed:
Fix rustdoc-map arg

jplatte commented on .cargo/config:
```diff
@@ -1,2 +1,5 @@
+[doc.extern-map.registries]
```
Tiny nit: Please move this section below `#[alias]`.

louisdewar replied:
Done

jplatte commented on .github/workflows/docs.yaml:
```diff
@@ -0,0 +1,29 @@
+on:
+  push:
+    branches: [main, docs-deployment]
```
Please remove `docs-deployment` (I'm guessing this was for testing?).

louisdewar replied:
Yep I can't run Github actions properly locally so I have to add my dev branch to test my changes, I'll make sure to remove it once I can confirm it's working well.

louisdewar replied:
It's removed now

jplatte reviewed (commented):
Two more small nits, then we can merge :)

louisdewar committed:
Remove dev branch and tidy up .cargo/config

louisdewar commented:
Great, looking at https://louisdewar.github.io/ruma/ruma/index.html the links to external API docs point to https://docs.rs and everything appear to be building fine

jplatte commented:
Thanks for taking care of this! :)

louisdewar commented:
No worries, I just wanted to try something easy to start out with before I start having a look at actual rust related issues
